File: bff-controle-escolar/src/student-access/student-access.service.ts
import { Injectable } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { firstValueFrom } from 'rxjs';
import { ConfigService } from '@nestjs/config';
import { StudentAccessDTO } from './dto/student-access.dto';

@Injectable()
export class StudentAccessService {
  private readonly studentAccessApiUrl: string;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService
  ) {
    this.studentAccessApiUrl = this.configService.get<string>('API_URL') ?? "http://localhost:8080/api";
  }

  async registerEntry(studentId: number): Promise<StudentAccessDTO> {
    const response = await firstValueFrom(
      this.httpService.post<StudentAccessDTO>(
        `${this.studentAccessApiUrl}/access/entry/${studentId}`,
        {}
      )
    );
    return response.data;
  }

  async registerExit(studentId: number): Promise<StudentAccessDTO> {
    const response = await firstValueFrom(
      this.httpService.post<StudentAccessDTO>(
        `${this.studentAccessApiUrl}/access/exit/${studentId}`,
        {}
      )
    );
    return response.data;
  }
}

File: backend/src/main/java/com/platformbuilders/controleescolar/studentaccess/service/StudentAccessService.java
package com.platformbuilders.controleescolar.studentaccess.service;

import com.platformbuilders.controleescolar.school.infra.repository.SchoolRepository;
import com.platformbuilders.controleescolar.student.domain.model.Student;
import com.platformbuilders.controleescolar.student.exception.StudentNotFoundException;
import com.platformbuilders.controleescolar.student.infra.repository.StudentRepository;
import com.platformbuilders.controleescolar.studentaccess.api.dto.StudentAccessDTO;
import com.platformbuilders.controleescolar.studentaccess.domain.model.StudentAccess;
import com.platformbuilders.controleescolar.studentaccess.infra.repository.StudentAccessRepository;
import io.swagger.v3.core.util.Json;
import jakarta.transaction.Transactional;
import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
@Slf4j
public class StudentAccessService {

    private final StudentAccessRepository accessRepository;
    private final StudentRepository studentRepository;
    private final ModelMapper modelMapper;

    @Autowired
    public StudentAccessService(
            StudentAccessRepository accessRepository,
            StudentRepository studentRepository,
            ModelMapper modelMapper) {
        this.accessRepository = accessRepository;
        this.studentRepository = studentRepository;
        this.modelMapper = modelMapper;
    }

    @Transactional
    public StudentAccessDTO registerEntry(Long studentId) {
        Student student = studentRepository.findById(studentId)
                .orElseThrow(StudentNotFoundException::new);

        StudentAccess access = new StudentAccess();
        access.setStudent(student);
        access.setEntryTime(LocalDateTime.now());
        log.info("Student {} entry registered at {}", studentId, LocalDateTime.now());

        StudentAccess savedAccess = accessRepository.save(access);
        return modelMapper.map(savedAccess, StudentAccessDTO.class);
    }

    @Transactional
    public StudentAccessDTO registerExit(Long studentId) {
        Student student = studentRepository.findById(studentId)
                .orElseThrow(StudentNotFoundException::new);

        StudentAccess access = accessRepository
                .findByStudentId(studentId)
                .orElseGet(() -> {
                    StudentAccess newAccess = new StudentAccess();
                    newAccess.setStudent(student);
                    newAccess.setEntryTime(LocalDateTime.now());
                    return newAccess;
                });

        access.setExitTime(LocalDateTime.now());
        log.info("Student {} exit registered at {}", studentId, LocalDateTime.now());
        StudentAccess savedAccess = accessRepository.save(access);
        return modelMapper.map(savedAccess, StudentAccessDTO.class);
    }

    public Page<StudentAccessDTO> findAll(Pageable pageable) {
        Page<StudentAccess> studentAccessPage = accessRepository.findAll(pageable);
        return studentAccessPage.map(access -> modelMapper.map(access, StudentAccessDTO.class));
    }
}